sourcegraph/sourcegraph
# batches: add organization setting to enable all members of org to become batch changes admins

Closes https://github.com/sourcegraph/sourcegraph/issues/50447

[Walkthrough](https://www.loom.com/share/895f3b1fe45d4d0a93cc4618ce51bfbf)

https://github.com/sourcegraph/accounts/issues/3 identified some areas where Batch Changes didn't solve some of their use cases. One is that Batch Changes created in an org namespace can only be operated by a site admin or the same user who created it.

This PR is a band-aid fix; the hope is that this is appropriately addressed when support for Batch Changes and teams is worked on.

I created a method in the `batches` service package to check for access to a batch change. The breakdown of the  logic of the method is as follows:

- If a Batch Change is in a user namespace, check if the current user is a site admin or the same user
- If a Batch Change is in an org namespace, then we check the organization settings for `orgs.allMembersBatchChangesAdmin` boolean field:
	- If false, we default to the existing check for if the current user is a site admin or the same user
	- If true, we check if the user belongs to the org or is a site admin or the same user

The following resolver operations are affected by this change:

| Operations |
|---|
|  MoveBatchChange |
|  CloseBatchChange |
|  DeleteBatchChange |
|  EnqueueChangesetSync |
|  ReenqueueChangeset |
| ApplyBatchChange |
|  CreateChangesetJobs (this is used for all bulk operations) |

The following resolver operations remain unchanged:

| Operations  |
|---|
| CreateEmptyBatchChange  |
| UpsertEmptyBatchChange  |
| CreateBatchSpec  |
| CreateBatchSpecFromRaw  |
| ExecuteBatchSpec  |
| CancelBatchSpec  |
| ReplaceBatchSpecInput  |
| UpsertBatchSpecInput  |

The logic for the above operations requires the current user to have access to the namespace regardless of where it was created (user or org namespace).

## Test plan

* `org.allMembersBatchChangesAdmin` enabled
	- You'll need to create an organization and navigate to the settings page for the organization [`/organizations/sourcegraph/settings`](https://sourcegraph.test:3443/organizations/sourcegraph/settings). 
	- In the settings page, set the value of `orgs.allMembersBatchChangesAdmin` to true (this will grant all members of the organization the ability to perform some administrative tasks on Batch Changes created in the org namespace)
	- Create a Batch Change in the organization namespace.
	- Access the Batch Change with another user account; this account shouldn't be a site admin but should be a member of the org you created.
	- This new account should be able to perform the following actions on the Batch Change created by the previous user in the org namespace:

		| Operations |
		|---|
		|  MoveBatchChange |
		|  CloseBatchChange |
		|  DeleteBatchChange |
		| ApplyBatchChange |
		|  Retry a changeset in failed state |
		|  All bulk operations |

* `org.allMembersBatchChangesAdmin` disabled or not set
	- You'll need to create an organization and navigate to the settings page for the organization [`/organizations/sourcegraph/settings`](https://sourcegraph.test:3443/organizations/sourcegraph/settings). 
	- Create a Batch Change in the organization namespace.
	- Access the Batch Change with another user account; this account shouldn't be a site admin but should be a member of the org you created.
	- The non-site admin account shouldn't be able to perform the following actions on Batch Changes created in the org namespace:

		| Operations |
		|---|
		|  MoveBatchChange |
		|  CloseBatchChange |
		|  DeleteBatchChange |
		| ApplyBatchChange |
		|  Retry a changeset in failed state |
		|  All bulk operations |
